rskj-core/src/main/java/co/rsk/peg/ABICallElection.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.peg;
+
+import org.ethereum.crypto.ECKey;
+
+import java.util.*;
+
+/**
+ * Representation of a given state of the election
+ * of an ABI function call by a series of known
+ * and authorized electors.
+ *
+ * @author Ariel Mendelzon
+ */
+public class ABICallElection {
+    private ABICallAuthorizer authorizer;
+    private Map<ABICallSpec, List<ABICallVoter>> votes;
+
+    public ABICallElection(ABICallAuthorizer authorizer, Map<ABICallSpec, List<ABICallVoter>> votes) {
+        this.authorizer = authorizer;
+        this.votes = votes;
+        validate();
+    }
+
+    public ABICallElection(ABICallAuthorizer authorizer) {
+        this.authorizer = authorizer;
+        this.votes = new HashMap<>();
+    }
+
+    public Map<ABICallSpec, List<ABICallVoter>> getVotes() {
+        return votes;
+    }
+
+    public void clear() {
+        this.votes = new HashMap<>();
+    }
+
+    /**
+     * Register voter's vote for callSpec
+     * @param callSpec the call spec the voter is voting for
+     * @param voter the voter's key
+     * @return whether the voting succeeded
+     */
+    public boolean vote(ABICallSpec callSpec, ABICallVoter voter) {
+        if (!authorizer.isAuthorized(voter))
+            return false;
+
+        if (!votes.containsKey(callSpec)) {
+            votes.put(callSpec, new ArrayList<>());
+        }
+
+        List<ABICallVoter> callVoters = votes.get(callSpec);
+
+        if (callVoters.contains(voter))
+            return false;
+
+        callVoters.add(voter);
+        return true;
+    }
+
+    /**
+     * Returns the election winner abi call spec, or null if there's none
+     * The vote authorizer determines the number of participants,
+     * whereas this class determines the number of votes that
+     * conforms a win
+     * @return the winner abi call spec
+     */
+    public ABICallSpec getWinner() {
+        for (Map.Entry<ABICallSpec, List<ABICallVoter>> specVotes : votes.entrySet()) {
```

Comment:
Isn't it possible for this to be two winners and being retrieved in different orders by different nodes?

Reply:
Yes it is. But I guess in that case it comes down to transaction execution order and mining.

But besides, it could only be more than one winner for an 'add public key' call, since votes are cleared after a 'create', 'commit' or 'rollback'.

Does that make sense?

Reply:
I saw how this is being used. I guess we can ignore this right now